### PR TITLE
SerialMapper for outer join

### PR DESCRIPTION
**Pull request by takkyuuplayer**
* mapping outer join rows

### CODE REVIEW

File: mapper.go [hunk 1]
```diff
@@ -256,6 +295,93 @@ func mapRowSerial(
...
+	}
+
+	colIndex = 0
+	for destIndex, dest := range destList {
+		if reflect.ValueOf(destVals[colIndex]).Elem().IsNil() { // *(Model.Field) == nil
+			colIndex += len(destList)
+			dest.Set(reflect.Zero(dest.Type()))
+			continue
+		}
+		model := reflect.New(destTypes[destIndex]) // *Model
```
Comment (keroxp) [inline mapper.go:363]: Does the original pointed value is always ignored? Isn't it possible to use original struct's fields if they exist?

```go
org := Value{}
orgPtr := &org
m.Map(rows, &orgPtr)
// It's desirable that &org == orgPtr even after mapping?
```

**Review by keroxp — CHANGES_REQUESTED** [review]

File: export_test.go [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package exql
+
+var ErrMapRowSerialDestination = errMapRowSerialDestination
```
[the remark this answers is not in the document]
Comment (takkyuuplayer) [reply export_test.go:3]: It is export_test pattern to use unexported variables in test

https://engineering.mercari.com/blog/entry/2018-08-08-080000/

File: mapper.go (hunk 1 above)
Comment (takkyuuplayer) [reply mapper.go:363]: It's possible but it means

* When `var orgPtr *Value` is used, always use `dest.Set(model)` 
* When `orgPtr := &Value{}` is used,
    * If data exists in the row, set each orgPtr.fields by  `f.Set`
    * If not exists, use `orgPtr.Set(nil value)`

I thought using `dest.Set(model)` in both cases was easier to read and maintain code.
What do you think? Do you think `&org == orgPtr even after mapping?` is more important in this case?

File: README.md [hunk 3]
```diff
@@ -217,6 +217,36 @@ func MapSerial() {
...
+		return
+	}
+	defer rows.Close()
+	serialMapper := exql.NewSerialMapper(func(i int) string {
+		// Each column's separator is `id`
+		return "id"
+	})
+	var users []*User
+	for rows.Next() {
+		var user *User              // Use *Model for outer join.
```
[the remark this answers is not in the document]
Comment (takkyuuplayer) [reply README.md]: updated in 85879c5

File: example/serial_mapper.go [hunk 4]
```diff
@@ -56,3 +56,33 @@ func MapSerial() {
 	}
 	// enumerate users...
 }
+
+func MapSerialLeftJoin() {
```
[the remark this answers is not in the document]
Comment (takkyuuplayer) [reply example/serial_mapper.go]: updated in 85879c5

File: mapper.go [hunk 5]
```diff
@@ -256,6 +295,93 @@ func mapRowSerial(
 	return row.Scan(destVals...)
 }
 
+// mapOuterJoinRowSerial maps rows to **Model. NULL value in ColumnSplitter column sets *Model = nil.
+// It can map inner join rows as well but using mapRowSerial is faster because less reflection logics.
+func mapOuterJoinRowSerial(
```
Comment (keroxp) [inline mapper.go:300]: I think both * and ** should be acceptable for inner/outer join in either case. There seems to be large part of duplicated codes between inner and outer. Could you migrate them into single function (within mapRowSerial)?

File: mapper.go (hunk 1 above)
Comment (keroxp) [reply mapper.go:363]: I felt it's opaque behavior that pointer value is swapped by reflection. I agree that in the 99% of cases, there are few reason to use pointer to existing value. So how about accepting only pointer of pointer to nil?

**Review by keroxp — CHANGES_REQUESTED** [review]

File: mapper.go (hunk 1 above)
Comment (takkyuuplayer) [reply mapper.go]: updated.

File: mapper.go (hunk 5 above)
Comment (takkyuuplayer) [reply mapper.go]: Updated. Now mapRowSerial accepts * and **, and I updated example, too.

**Review by keroxp — APPROVED** [review]
Great Patch 👍  That's a big step for our journey!

**Pull request merged**